1.x: java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 329648 free bytes and 321KB until OOM; failed due to fragmentation...

**Trying to sync ~380,000 documents with no attachments**

App keeps crashing. I think it synced about ~150,000 and now it crashes every 10 seconds. Tried restarting the device, and force closing app. 

Cordova/Ionic App
Current App Size: 2.71 GB

Device: Asus Tablet P00A
Total Memory: 1.9 GB
Android: 6.0

**Is there a limit to how many documents a device can handle?**

```
Uncaught exception thrown by finalizer
10-14 11:59:26.762 6086-6094/? E/System: java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 329648 free bytes and 321KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)
      at com.android.internal.os.BinderInternal$GcWatcher.finalize(BinderInternal.java:53)
      at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:228)
      at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:206)
      at java.lang.Thread.run(Thread.java:818)

10-14 11:59:26.762 6086-6094/? I/System: FinalizerDaemon: finalize objects = 26
10-14 11:59:26.792 6086-6181/? I/art: Alloc concurrent mark sweep GC freed 774(51KB) AllocSpace objects, 0(0B) LOS objects, 37% free, 6MB/10MB, paused 1.365ms total 31.195ms
10-14 11:59:26.793 6086-6177/? I/art: WaitForGcToComplete blocked for 724.833ms for cause Alloc
10-14 11:59:26.793 6086-6177/? I/art: Starting a blocking GC Alloc
10-14 11:59:26.793 6086-6096/? I/art: WaitForGcToComplete blocked for 9.500s for cause HeapTrim
10-14 11:59:26.793 6086-6095/? I/art: Waiting for a blocking GC Alloc
10-14 11:59:26.793 6086-6212/? I/art: Waiting for a blocking GC Alloc
10-14 11:59:26.793 6086-6203/? I/art: Waiting for a blocking GC Alloc

10-14 11:59:26.793 6086-6177/? W/Router: Router unable to route request to do_GET_Document_changes
  java.lang.reflect.InvocationTargetException
      at java.lang.reflect.Method.invoke(Native Method)
      at com.couchbase.lite.router.Router.start(Router.java:556)
      at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
      at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
      at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
      at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
      at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
      at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
      at java.lang.Thread.run(Thread.java:818)
   Caused by: java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack trace available
10-14 11:59:26.793 6086-6183/? I/art: Waiting for a blocking GC Alloc
```

I was able to sync with the same # and type of documents on a Nexus 6 Running 7.0 without any crash.

Total Sync time: 4.28 hours
App Data Size: 9.74 GB

However, now the app continues to crash after ~1 min.

There are no views running on app launch. Could this still be related to the large documents?

Similar memory issue. 

App launch 1

```
 Throwing OutOfMemoryError "Failed to allocate a 960400 byte allocation with 620104 free bytes and 605KB until OOM"
10-15 10:15:08.211 27472-27589/com.xxx.xxxW/Router: Router unable to route request to do_GET_Document_changes
                                                                     java.lang.reflect.InvocationTargetException
                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                         at com.couchbase.lite.router.Router.start(Router.java:556)
                                                                         at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
                                                                         at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
                                                                         at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
                                                                         at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
                                                                         at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
                                                                         at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
                                                                         at java.lang.Thread.run(Thread.java:761)
                                                                      Caused by: java.lang.OutOfMemoryError: Failed to allocate a 960400 byte allocation with 620104 free bytes and 605KB until OOM
                                                                         at java.util.Arrays.copyOf(Arrays.java:3231)
                                                                         at java.util.Arrays.copyOf(Arrays.java:3204)
                                                                         at java.util.ArrayList.grow(ArrayList.java:249)
                                                                         at java.util.ArrayList.ensureExplicitCapacity(ArrayList.java:223)
                                                                         at java.util.ArrayList.ensureCapacityInternal(ArrayList.java:215)
                                                                         at java.util.ArrayList.add(ArrayList.java:441)
                                                                         at com.couchbase.lite.router.Router.responseBodyForChanges(Router.java:1529)
                                                                         at com.couchbase.lite.router.Router.doChanges(Router.java:1846)
                                                                         at com.couchbase.lite.router.Router.do_GET_Document_changes(Router.java:1730)
                                                                         at java.lang.reflect.Method.invoke(Native Method) 
                                                                         at com.couchbase.lite.router.Router.start(Router.java:556) 
                                                                         at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141) 
                                                                         at javax.servlet.http.HttpServlet.service(HttpServlet.java:802) 
                                                                         at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251) 
                                                                         at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176) 
                                                                         at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988) 
                                                                         at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223) 
                                                                         at java.lang.Thread.run(Thread.java:761) 
10-15 10:15:08.424 27472-27680/com.xxx.xxx W/art: Suspending all threads took: 189.607ms
10-15 10:15:08.692 27472-27608/com.xxx.xxx W/art: Suspending all threads took: 203.904ms
10-15 10:15:08.761 27472-28233/com.xxx.xxx W/art: Suspending all threads took: 38.828ms
10-15 10:15:08.815 27472-27589/com.xxx.xxx W/System.err: [Sat Oct 15 10:15:08 CDT 2016] TJWS: IO error: java.net.SocketException: Broken pipe in processing a request from /::1:5984 / java.net.Socket
10-15 10:15:08.828 27472-27479/com.xxx.xxx W/art: Suspending all threads took: 25.273ms
```
App Launch 2

```
10-15 10:16:12.165 31400-31505/com.xxx.xxx E/UCI4C: u_getDataDirectory()=/data/user/0/com.xxx.xxx/files/icu
10-15 10:16:12.250 31400-31505/com.xxx.xxx W/Router: Error updating doc: _design/todo
10-15 10:16:13.303 31400-31532/com.xxx.xxx W/Sync: Error converting lastSequence: null to long.  Using 0
10-15 10:16:15.820 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 24.471ms
10-15 10:16:19.870 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 62.690ms
...
10-15 10:16:51.267 31400-31537/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 8012 byte allocation with 476600 free bytes and 465KB until OOM; failed due to fragmentation (required continguous free 8192 bytes where largest contiguous free 0 bytes)"
10-15 10:16:53.567 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 549.546ms
10-15 10:16:53.669 31400-31797/com.xxx.xxx W/art: Suspending all threads took: 101.203ms
10-15 10:16:53.678 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 6.869ms
10-15 10:16:55.765 31400-31797/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 8204 byte allocation with 465544 free bytes and 454KB until OOM; failed due to fragmentation (required continguous free 12288 bytes where largest contiguous free 0 bytes)"
10-15 10:16:58.193 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 581.951ms
...
10-15 10:16:59.356 31400-31535/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 4184 byte allocation with 460056 free bytes and 449KB until OOM; failed due to fragmentation (required continguous free 8192 bytes where largest contiguous free 0 bytes)"
10-15 10:16:59.362 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 5.441ms
10-15 10:17:00.158 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 674.234ms
10-15 10:17:00.233 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 11.899ms
...
10-15 10:17:02.780 31400-31714/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 128 byte allocation with 467976 free bytes and 457KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)"
10-15 10:17:03.318 31400-31712/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 8012 byte allocation with 418408 free bytes and 408KB until OOM; failed due to fragmentation (required continguous free 8192 bytes where largest contiguous free 0 bytes)"
10-15 10:17:03.989 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 621.539ms
10-15 10:17:04.087 31400-31797/com.xxx.xxx W/art: Suspending all threads took: 98.545ms
10-15 10:17:04.096 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 7.560ms
10-15 10:17:04.242 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 12.812ms
10-15 10:17:04.309 31400-31787/com.xxx.xxx W/art: Suspending all threads took: 66.667ms
...
10-15 10:17:08.655 31400-31797/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 128 byte allocation with 452080 free bytes and 441KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)"
10-15 10:17:09.319 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 575.775ms
...
                                                                  
                                                                  [ 10-15 10:17:10.444   270:  270 W/         ]
                                                                  debuggerd: resuming target 402
                                                                  
                                                                  [ 10-15 10:17:10.444   270:  270 W/         ]
                                                                  debuggerd: handling request: pid=403 uid=1047 gid=1005 tid=403
                                                                  
                                                                  
                                                                  [ 10-15 10:17:10.507   270:  270 W/         ]
                                                                  debuggerd: resuming target 403
                                                                  
                                                                  [ 10-15 10:17:10.508   270:  270 W/         ]
                                                                  debuggerd: handling request: pid=404 uid=1019 gid=1019 tid=404
10-15 10:17:11.188 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 568.395ms
10-15 10:17:11.274 31400-31400/com.xxx.xxx W/art: Suspending all threads took: 85.485ms
10-15 10:17:11.281 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 5.289ms
10-15 10:17:11.350 31400-31400/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 436 byte allocation with 463288 free bytes and 452KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)"
10-15 10:17:11.358 31400-31400/com.xxx.xxx E/InputEventReceiver: Exception dispatching input event.
10-15 10:17:11.362 31400-31400/com.xxx.xxx E/MessageQueue-JNI: Exception in MessageQueue callback: handleReceiveCallback
10-15 10:17:11.927 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 501.765ms
                                                                  
                                                                  [ 10-15 10:17:12.011   270:  270 W/         ]
                                                                  debuggerd: resuming target 8602
10-15 10:17:12.062 31400-31841/com.xxx.xxx W/art: Suspending all threads took: 64.376ms
10-15 10:17:12.083 31400-31841/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 128 byte allocation with 446400 free bytes and 435KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)"
10-15 10:17:12.709 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 281.941ms
10-15 10:17:12.816 31400-31714/com.xxx.xxx W/art: Suspending all threads took: 107.423ms
...
10-15 10:17:20.835 31400-31841/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 448 byte allocation with 468936 free bytes and 457KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)"
10-15 10:17:21.432 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 487.170ms
10-15 10:17:21.503 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 7.885ms
10-15 10:17:21.603 31400-31570/com.xxx.xxx W/art: Suspending all threads took: 99.574ms
10-15 10:17:22.155 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 208.662ms
10-15 10:17:22.791 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 343.644ms
10-15 10:17:22.879 31400-31400/com.xxx.xxx W/art: Suspending all threads took: 88.918ms
10-15 10:17:22.890 31400-31400/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 40 byte allocation with 468176 free bytes and 457KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)"
10-15 10:17:22.890 31400-31400/com.xxx.xxx E/MessageQueue-JNI: java.lang.OutOfMemoryError: Failed to allocate a 436 byte allocation with 463288 free bytes and 452KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)
10-15 10:17:23.463 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 515.730ms
10-15 10:17:23.562 31400-31411/com.xxx.xxx W/art: Suspending all threads took: 99.140ms
10-15 10:17:23.570 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 5.674ms
10-15 10:17:24.242 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 293.823ms
...
10-15 10:17:28.037 31400-31570/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 37 byte allocation with 466056 free bytes and 455KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)"
10-15 10:17:28.732 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 273.208ms
10-15 10:17:28.835 31400-31787/com.xxx.xxx W/art: Suspending all threads took: 102.701ms
10-15 10:17:29.527 31400-31409/com.xxx.xxx W/art: Suspending all threads took: 568.427ms
...
10-15 10:17:34.232 31400-31841/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 2072 byte allocation with 465936 free bytes and 455KB until OOM; failed due to fragmentation (required continguous free 4096 bytes where largest contiguous free 0 bytes)"

```
App launch 3
```

10-15 10:03:51.359 20676-20892/com.xxx.xxx I/art: WaitForGcToComplete blocked for 2.860s for cause Alloc
10-15 10:03:51.360 20676-20892/com.xxx.xxx I/art: Starting a blocking GC Alloc
10-15 10:03:51.360 20676-20882/com.xxx.xxx I/art: Waiting for a blocking GC Alloc
10-15 10:03:51.360 20676-20683/com.xxx.xxx I/art: Waiting for a blocking GC Alloc
10-15 10:03:51.360 20676-20712/com.xxx.xxx I/art: Waiting for a blocking GC Alloc
10-15 10:03:51.360 20676-20885/com.xxx.xxx I/art: Waiting for a blocking GC Alloc
10-15 10:03:51.360 20676-20676/com.xxx.xxx I/art: Waiting for a blocking GC Alloc
10-15 10:03:51.422 20676-20848/com.xxx.xxx W/art: Throwing OutOfMemoryError "Failed to allocate a 68 byte allocation with 273456 free bytes and 267KB until OOM; failed due to fragmentation (required continguous free 4096 bytes for a new buffer where largest contiguous free 0 bytes)" (recursive case)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art: "Acme.Utils.ThreadPool(1)-PooledThread: Acme.Serve.Serve$ServeConnection@3d6c1d1" daemon prio=5 tid=39 Runnable
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   | group="TJWS threads" sCount=0 dsCount=0 obj=0x12c13280 self=0xa5d07b00
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   | sysTid=20848 nice=0 cgrp=default sched=0/0 handle=0x7a31f920
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   | state=R schedstat=( 51615044345 4183253664 34357 ) utm=4955 stm=205 core=2 HZ=100
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   | stack=0x7a21d000-0x7a21f000 stackSize=1038KB
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   | held mutexes= "mutator lock"(shared held)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #00 pc 0034dced  /system/lib/libart.so (_ZN3art15DumpNativeStackERNSt3__113basic_ostreamIcNS0_11char_traitsIcEEEEiP12BacktraceMapPKcPNS_9ArtMethodEPv+128)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #01 pc 0032e569  /system/lib/libart.so (_ZNK3art6Thread9DumpStackERNSt3__113basic_ostreamIcNS1_11char_traitsIcEEEEbP12BacktraceMap+304)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #02 pc 0032c7bb  /system/lib/libart.so (_ZN3art6Thread21ThrowOutOfMemoryErrorEPKc+174)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #03 pc 0018834d  /system/lib/libart.so (_ZN3art2gc4Heap21ThrowOutOfMemoryErrorEPNS_6ThreadEjNS0_13AllocatorTypeE+836)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #04 pc 00190839  /system/lib/libart.so (_ZN3art2gc4Heap22AllocateInternalWithGcEPNS_6ThreadENS0_13AllocatorTypeEbjPjS5_S5_PPNS_6mirror5ClassE+21376)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #05 pc 000ffb53  /system/lib/libart.so (_ZN3art2gc4Heap24AllocObjectWithAllocatorILb1ELb1ENS_6mirror16SetLengthVisitorEEEPNS3_6ObjectEPNS_6ThreadEPNS3_5ClassEjNS0_13AllocatorTypeERKT1_+2358)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #06 pc 000ff179  /system/lib/libart.so (_ZN3art6mirror11ObjectArrayINS0_17StackTraceElementEE5AllocEPNS_6ThreadEPNS0_5ClassEiNS_2gc13AllocatorTypeE+64)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #07 pc 00329e67  /system/lib/libart.so (_ZN3art30BuildInternalStackTraceVisitorILb0EE4InitEi+106)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #08 pc 00329dab  /system/lib/libart.so (_ZNK3art6Thread24CreateInternalStackTraceILb0EEEP8_jobjectRKNS_33ScopedObjectAccessAlreadyRunnableE+118)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #09 pc 002b22cd  /system/lib/libart.so (_ZN3artL32Throwable_nativeFillInStackTraceEP7_JNIEnvP7_jclass+28)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:   native: #10 pc 00015e3d  /system/framework/arm/boot.oat (Java_java_lang_Throwable_nativeFillInStackTrace__+72)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.Throwable.nativeFillInStackTrace!(Native method)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.Throwable.fillInStackTrace(Throwable.java:774)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.Throwable.<init>(Throwable.java:257)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.Error.<init>(Error.java:70)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.VirtualMachineError.<init>(VirtualMachineError.java:53)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.OutOfMemoryError.<init>(OutOfMemoryError.java:58)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at com.couchbase.lite.router.Router.changesDictForRevision(Router.java:1505)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at com.couchbase.lite.router.Router.responseBodyForChanges(Router.java:1528)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at com.couchbase.lite.router.Router.doChanges(Router.java:1846)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at com.couchbase.lite.router.Router.do_GET_Document_changes(Router.java:1730)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.reflect.Method.invoke!(Native method)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at com.couchbase.lite.router.Router.start(Router.java:556)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
10-15 10:03:51.451 20676-20848/com.xxx.xxx W/art:     at java.lang.Thread.run(Thread.java:761)

```

--- a/src/main/java/com/couchbase/lite/router/Router.java
+++ b/src/main/java/com/couchbase/lite/router/Router.java
@@ -88,6 +88,10 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
 
     private static final String CONTENT_TYPE_JSON = "application/json";
 
+    // Android cannot hold unlimited changes data.
+    // See ChangesOptions.limit default value (Integer.MAX_VALUE)
+    public static final int DEF_CHANGES_LIMIT = 10000;
+
     /**
      * Options for what metadata to include in document bodies
      */
@@ -1557,7 +1561,7 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
         return result;
     }
 
-    private Map<String, Object> responseBodyForChangesWithConflicts(List<RevisionInternal> changes, long since) {
+    private Map<String, Object> responseBodyForChangesWithConflicts(List<RevisionInternal> changes, long since, int limit) {
         // Assumes the changes are grouped by docID so that conflicts will be adjacent.
         List<Map<String, Object>> entries = new ArrayList<Map<String, Object>>();
         String lastDocID = null;
@@ -1581,6 +1585,9 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
                 return Misc.SequenceCompare((Long) e1.get("seq"), (Long) e2.get("seq"));
             }
         });
+
+        if (entries.size() > limit)
+            entries = entries.subList(0, limit);
 
         Long lastSeq;
         if (entries.size() == 0) {
@@ -1792,8 +1799,8 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
         options.setIncludeDocs(changesIncludesDocs);
         options.setIncludeConflicts(changesIncludesConflicts);
         options.setSortBySequence(!options.isIncludeConflicts());
-        // TODO: descending option is not supported by ChangesOptions
-        options.setLimit(getIntQuery("limit", options.getLimit()));
+        options.setLimit(getIntQuery("limit", DEF_CHANGES_LIMIT)); // def value was options.getLimit().
+        // NOTE: descending option is not supported by ChangesOptions
 
         int since = getIntQuery("since", 0);
         timeoutLastSeqence = since;
@@ -1878,7 +1885,7 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
             return new Status(0);
         } else {
             if (options.isIncludeConflicts()) {
-                connection.setResponseBody(new Body(responseBodyForChangesWithConflicts(changes, since)));
+                connection.setResponseBody(new Body(responseBodyForChangesWithConflicts(changes, since, options.getLimit())));
             } else {
                 connection.setResponseBody(new Body(responseBodyForChanges(changes, since)));
             }

--- a/src/main/java/com/couchbase/lite/router/Router.java
+++ b/src/main/java/com/couchbase/lite/router/Router.java
@@ -88,10 +88,6 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
 
     private static final String CONTENT_TYPE_JSON = "application/json";
 
-    // Android cannot hold unlimited changes data.
-    // See ChangesOptions.limit default value (Integer.MAX_VALUE)
-    public static final int DEF_CHANGES_LIMIT = 10000;
-
     /**
      * Options for what metadata to include in document bodies
      */
@@ -1799,7 +1795,7 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
         options.setIncludeDocs(changesIncludesDocs);
         options.setIncludeConflicts(changesIncludesConflicts);
         options.setSortBySequence(!options.isIncludeConflicts());
-        options.setLimit(getIntQuery("limit", DEF_CHANGES_LIMIT)); // def value was options.getLimit().
+        options.setLimit(getIntQuery("limit", options.getLimit()));
         // NOTE: descending option is not supported by ChangesOptions
 
         int since = getIntQuery("since", 0);

